### PR TITLE
Added type option to SetSelection

### PR DESCRIPTION
**Pull request by ghDaYuYu**
Hi TheQwertiest
I have just started coding with fb2k api and I am trying some customization on a library view component.
In the SMP repo I can see that the SetSelection type has a fixed 'undefined' value.
I dont know, there may be reasons for not changing that value but I tested it and the added parameter seams to work fine and be useful.
Hope that you find it useful too and accept the pull!

Best regards

### CODE REVIEW

**Comment by TheQwertiest** [comment]
Thanks for the contribution!
Still, I have to ask: what's the use-case for this feature?

**Comment by ghDaYuYu** [comment]
Good night TheQwertiest, sorry for taking some time to respond,
here is a simple use case that comes to my mind : )
 
component CA and component CB share selections (to display different views of the same item) by means of 
selection holder callbacks. when CA makes a selection, CB syncs to it and the other way around
.
This is true when both components are operating in the same source mode.

component CA operates either as library viewer or playlist browser and its container is an SMP panel.
CA sets selections of either the library viewer or playlist browser type, and listens to selection types 
according to its operating mode.

component CB operating mode is playlist browser, its container is an ordinary fb2k panel
CB sets selections of the playlist browser type and syncs to that type when selections are available.

The added parameter to SetSelection when set by CA is used by CB to either sync to or discard CA selections.
When CB sets its selection type, CA use type info to sync or discard depending of its operation mode.

Developers can decide on either to ignore or accept undefined selections or to add a switchable optional property to define that behaviour. Selection types checks, as you known, were already available for SMP panel by means of fb.GetSelectionType() - for non SMP Panel originated selections. Now is also possible to set that value by means of selectionHolder.SetSelection(metalist, type) 

In sort : ) the type parameter provides SMP panel components the ability to broadcast its type of selection to non SMP panel components, also to other SMP panel components as an alternative to on_notify_data().

Hope this was ok as use case, not sure if this was what your asked for, tell me if is not... also have some code examples at hand I can prepare it if you thing it could be suitable for documentation.

Thanks in advance, hope you decide to incorporate this pull to your repo and will come quite handy on the component mod I am working on.
Good Night

Regards

File: foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp [hunk 1]
```diff
@@ -89,11 +89,40 @@ void JsFbUiSelectionHolder::SetPlaylistTracking()
...
-void JsFbUiSelectionHolder::SetSelection( JsFbMetadbHandleList* handles )
+const std::array<const GUID*, 7> guids = {
+    &contextmenu_item::caller_undefined,
+    &contextmenu_item::caller_active_playlist_selection,
+    &contextmenu_item::caller_active_playlist,
+    &contextmenu_item::caller_playlist_manager,
+    &contextmenu_item::caller_now_playing,
+    &contextmenu_item::caller_keyboard_shortcut_list,
+    &contextmenu_item::caller_media_library_viewer
+};
```
Comment (TheQwertiest) [inline foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp:100]: This should be wrapped into a function that checks the index (see https://github.com/TheQwertiest/foo_spider_monkey_panel/blob/ab822e6508cd0861116525fcf68856c778f34709/foo_spider_monkey_panel/utils/art_helpers.cpp#L226 for example) and moved to the anonymous namespace.

File: foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp [hunk 2]
```diff
@@ -89,11 +89,40 @@ void JsFbUiSelectionHolder::SetPlaylistTracking()
...
+{
+    SmpException::ExpectTrue( handles, "handles argument is null" );
+
+    switch ( optArgCount )
+    {
+    case 0:
+        holder_->set_selection_ex( handles->GetHandleList(), *guids.at( type ) );
+        break;
+    case 1:
+        optArgCount = optArgCount;
```
Comment (TheQwertiest) [inline foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp:120]:
```suggestion
```

File: foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.h [hunk 3]
```diff
@@ -36,7 +36,8 @@ class JsFbUiSelectionHolder
 public:
     void SetPlaylistSelectionTracking();
     void SetPlaylistTracking();
-    void SetSelection( JsFbMetadbHandleList* handles );
+    void SetSelection( JsFbMetadbHandleList* handles, uint8_t type);
```
Comment (TheQwertiest) [inline foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.h:39]:
```suggestion
    void SetSelection( JsFbMetadbHandleList* handles, uint8_t type = 0);
```

File: foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp [hunk 4]
```diff
@@ -89,11 +89,40 @@ void JsFbUiSelectionHolder::SetPlaylistTracking()
...
+}
+
+void JsFbUiSelectionHolder::SetSelectionWithOpt( size_t optArgCount, JsFbMetadbHandleList* handles, uint8_t type )
+{
+    SmpException::ExpectTrue( handles, "handles argument is null" );
+
+    switch ( optArgCount )
+    {
+    case 0:
+        holder_->set_selection_ex( handles->GetHandleList(), *guids.at( type ) );
```
Comment (TheQwertiest) [inline foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp:117]:
```suggestion
        SetSelection( handles, type );
```

File: foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp [hunk 5]
```diff
@@ -89,11 +89,40 @@ void JsFbUiSelectionHolder::SetPlaylistTracking()
...
+        break;
+    case 1:
+        optArgCount = optArgCount;
+        holder_->set_selection_ex( handles->GetHandleList(), contextmenu_item::caller_undefined );
```
Comment (TheQwertiest) [inline foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp:121]:
```suggestion
       SetSelection( handles );
```

File: foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp [hunk 6]
```diff
@@ -89,11 +89,40 @@ void JsFbUiSelectionHolder::SetPlaylistTracking()
...
     SmpException::ExpectTrue( handles, "handles argument is null" );
 
-    holder_->set_selection( handles->GetHandleList() );
+    holder_->set_selection_ex( handles->GetHandleList(), *guids.at( type ) );
+}
+
+void JsFbUiSelectionHolder::SetSelectionWithOpt( size_t optArgCount, JsFbMetadbHandleList* handles, uint8_t type )
+{
+    SmpException::ExpectTrue( handles, "handles argument is null" );
+
```
Comment (TheQwertiest) [inline foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp:113]:
```suggestion
```

**Comment by TheQwertiest** [comment]
Another thing: docs are needed to be altered accordingly - https://github.com/TheQwertiest/foo_spider_monkey_panel/blob/b01c5de11367910d9761fc248fc8aebe8638e741/component/docs/js/foo_spider_monkey_panel.js#L2773

File: foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp (hunk 1 above)
Comment (ghDaYuYu) [reply foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp:100]: also though about it, did not like it neither
there is another reference to that array here:

https://github.com/TheQwertiest/foo_spider_monkey_panel/blob/ab822e6508cd0861116525fcf68856c778f34709/foo_spider_monkey_panel/js_objects/fb_utils.cpp#L552-L570

I wonder where would that function be placed so both functions can call it... utils\menu_helpers.h ?

https://github.com/TheQwertiest/foo_spider_monkey_panel/blob/ab822e6508cd0861116525fcf68856c778f34709/foo_spider_monkey_panel/utils/menu_helpers.h#L1

The namespace there is smp::utils

what would you recommend? Just wrap it before SetSelection() or make it available to GetSelection()

Comment (TheQwertiest) [reply foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp:100]: Nice catch =)

> I wonder where would that function be placed so both functions can call it... utils\menu_helpers.h ?

Just create new files in the root, i.e. `\foo_spider_monkey_panel\foo_spider_monkey_panel\context_menu_caller.(h/cpp)` with `namespace smp`. I have plans to reorganize some source files anyway. So will clean it up later.

Comment (ghDaYuYu) [reply foo_spider_monkey_panel/js_objects/fb_ui_selection_holder.cpp:100]: Well, oops...
The whole block might be as well be replaced by (my bad, sorry)
const GUID type = ui_selection_manager_v2::get()->get_selection_type( 0 );
and can maybe let GetSelectionType() alone as its doing the opposite conversion

I'll revise after having some rest. Also will check the changes in the documentation...
Thanks 
Best regards

**Comment by TheQwertiest** [comment]
Thanks for the contribution and sorry for the delay!

I've rebased and merged your PR in https://github.com/TheQwertiest/foo_spider_monkey_panel/pull/128 (since I didn't have required permissions to push to your fork).

**Pull request closed without merging**